packages/hooks/src/createUseStorageState/index.ts

```diff
@@ -67,17 +72,58 @@ export function createUseStorageState(getStorage: () => Storage | undefined) {
       const currentState = isFunction(value) ? value(state) : value;
       setState(currentState);
 
-      if (isUndef(currentState)) {
-        storage?.removeItem(key);
-      } else {
-        try {
-          storage?.setItem(key, serializer(currentState));
-        } catch (e) {
-          console.error(e);
+      try {
+        let newValue: string | null;
+        const oldValue = storage?.getItem(key);
+
+        if (isUndef(currentState)) {
+          newValue = null;
+          storage?.removeItem(key);
+        } else {
+          newValue = serializer(currentState);
+          storage?.setItem(key, newValue);
         }
+
+        dispatchEvent(
+          // send custom event to communicate within same page
+          // importantly this should not be a StorageEvent since those cannot
+          // be constructed with a non-built-in storage area
+          new CustomEvent(SYNC_STORAGE_EVENT_NAME, {
+            detail: {
+              key,
+              newValue,
+              oldValue,
+              storageArea: storage,
+            },
+          }),
+        );
+      } catch (e) {
+        onError(e);
       }
     };
 
+    const syncState = (event: StorageEvent) => {
+      if (event.key !== key || event.storageArea !== storage) {
+        return;
+      }
+
+      setState(getStoredValue());
+    };
+
+    const syncStateFromCustomEvent = (event: CustomEvent<StorageEvent>) => {
+      syncState(event.detail);
+    };
+
+    // from another document
+    useEventListener('storage', syncState, {
+      enable: listenStorageChange,
+    });
+
+    // from the same document but different hooks
+    useEventListener(SYNC_STORAGE_EVENT_NAME, syncStateFromCustomEvent, {
```

Comment:
`syncStateFromCustomEvent` will be triggered by its own `updateState` function, causing the `setState` method to be called twice.

<img width="458" alt="image" src="https://github.com/alibaba/hooks/assets/20217146/d95bd1ed-c273-4d6a-a8bb-cf9674410857">

Reply:
because its own listener also detect its own trigger. whether can we use `const hookId = useRef(Symbol(xx))` as current hook's identify. when calling  `dispatchEvent` , pass the `hookId`. In `syncStateFromCustomEvent`, compare hookId.
sorry, English is poor.

Reply:
@bowencool Thank you, I will check it again and fix.

Reply:
I believe that these codes can help:
```diff
  const updateState = (value?: SetState<T>) => {
      const currentState = isFunction(value) ? value(state) : value;
+    if (!listenStorageChange) {
        setState(currentState);
+   }
  ...
```